### PR TITLE
Feature camel ectd3 - Etcd3AggregationRepository using jetcd

### PR DESCRIPTION
**Pull request by malogan82**
[ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
[ ] Each commit in the pull request should have a meaningful subject line and body.
[ ] If you're unsure, you can format the pull request title like `[CAMEL-XXX] Fixes bug in camel-file component`, where you replace `CAMEL-XXX` with the appropriate JIRA issue.
[ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
[ ] Run `mvn clean install -Psourcecheck` in your module with source check enabled to make sure basic checks pass and there are no checkstyle violations. A more thorough check will be performed on your pull request automatically.
Below are the contribution guidelines:
https://github.com/apache/camel/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by lburgazzoli** [comment]
since the camel-etcd component does does not support etcd v3, I'd recommend to create a new [camel-etcd3]( https://issues.apache.org/jira/browse/CAMEL-10173) component and this code there.

**Comment by malogan82** [comment]
ok thanks

File: components/camel-etcd3/pom.xml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    <dependencies>
+
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-cloud</artifactId>
+        </dependency>
+
+        <!-- etcd -->
+        <dependency>
+            <groupId>org.mousio</groupId>
```
Comment (oscerd) [inline components/camel-etcd3/pom.xml:44]: I don't think etcd4j is needed

Comment (malogan82) [reply components/camel-etcd3/pom.xml:44]: right, I remove this dependency

**Comment by oscerd** [comment]
Thanks

**Comment by malogan82** [comment]
ok thanks

**Comment by oscerd** [comment]
And Please can you amend the commit messages translating to English? Thanks

**Comment by malogan82** [comment]
sure

**Comment by malogan82** [comment]
I've created another component camel-redis, creating the class RedisAggregationRepository which uses redisson library, I don't know if I'm wrong, but maybe it can be useful to apply aggregator eip using redis

**Comment by malogan82** [comment]
@oscerd ok I create a new pull request for camel-reply thanks

File: parent/pom.xml [hunk 2]
```diff
@@ -2021,6 +2027,11 @@
         <groupId>org.apache.camel</groupId>
         <artifactId>camel-reactor</artifactId>
         <version>${project.version}</version>
+      </dependency>
+	  <dependency>
+        <groupId>org.apache.camel</groupId>
```
Comment (oscerd) [inline parent/pom.xml:2032]: I think camel-redis shouldn't be here

**Comment by lburgazzoli** [comment]
@malogan82 can you also please squash the commits ?

**Comment by malogan82** [comment]
@lburgazzoli I've squashed the commits

File: camel-dependencies/pom.xml [hunk 3]
```diff
@@ -486,6 +487,7 @@
     <rdf4j-rio-version>2.4.4</rdf4j-rio-version>
     <reactive-streams-version>1.0.3</reactive-streams-version>
     <reactor-version>3.2.16.RELEASE</reactor-version>
+	<redisson-version>3.13.2</redisson-version>
```
Comment (lburgazzoli) [inline camel-dependencies/pom.xml:490]: does not seem to be related

File: camel-dependencies/pom.xml [hunk 4]
```diff
@@ -332,6 +332,7 @@
     <jcr-version>2.0</jcr-version>
     <jedis-client-version>3.1.0</jedis-client-version>
     <jersey-version>2.28</jersey-version>
+	<jetcd-version>0.5.3</jetcd-version>
```
Comment (lburgazzoli) [inline camel-dependencies/pom.xml:335]: looks like this and other xml files have wrong indentation

File: components/camel-etcd3/pom.xml [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        
+        <dependency>
+            <groupId>io.etcd</groupId>
+            <artifactId>jetcd-all</artifactId>
+            <version>${jetcd-version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>io.etcd</groupId>
+            <artifactId>jetcd-core</artifactId>
```
Comment (lburgazzoli) [inline components/camel-etcd3/pom.xml:57]: either you use `jetcd-all` or you use `jetcd-core` as `jetcd-all` shades all the `jetcd` artefacts and dependencies

File: components/camel-etcd3/pom.xml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-http</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-jetty</artifactId>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (lburgazzoli) [inline components/camel-etcd3/pom.xml:89]: Those dependencies need to be leaned up and you also need to add some test case for the aggregation repository

File: components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+	public Exchange add(CamelContext camelContext, String key, Exchange exchange) {
+		if (optimistic) {
+			throw new UnsupportedOperationException();
+		}
+		LOG.trace("Adding an Exchange with ID {} for key {} in a thread-safe manner.", exchange.getExchangeId(), key);
+		Lock lock = null;
+		DefaultExchangeHolder newHolder = DefaultExchangeHolder.marshal(exchange, true, allowSerializedHeaders);
+		try {
+			lock = client.getLockClient();
+			lock.lock(ByteSequence.from(key.getBytes()), 60000);
```
Comment (lburgazzoli) [inline components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java:177]: the lease id need to be obtained from the `LeaseClient` and not hard-coded, however the `lock` APIs may not behave as you'd expect, see [notes-on-the-usage-of-lock-and-lease](https://github.com/etcd-io/etcd/blob/master/Documentation/learning/why.md#notes-on-the-usage-of-lock-and-lease).

It may be better to use transactions as you can specify the conditions to be met when performing the operation.

File: components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+			throw new IllegalArgumentException("Recovery interval must be zero or a positive integer.");
+		}
+		StringHelper.notEmpty(prefixName, "prefixName");
+		client = Client.builder().endpoints(endpoint).build();
+		kvClient = client.getKVClient();
+	}
+
+	@Override
+	protected void doStop() throws Exception {
+		client.close();
```
Comment (lburgazzoli) [inline components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java:448]: Nee to check if client is `null`

File: components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+			scanned = Collections.unmodifiableSet(keys);
+		} catch (InterruptedException | ExecutionException e) {
+			LOG.error(e.getMessage(), e);
+		}
+		return scanned;
+	}
+
+	@Override
+	protected void doStart() throws Exception {
+		if (maximumRedeliveries < 0) {
```
Comment (lburgazzoli) [inline components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java:435]: Validation can be moved to `onInit` to check correctness of the parameters early

File: components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+					key);
+			if (useRecovery) {
+				LOG.trace(
+						"Putting an exchange with ID {} for key {} into a recoverable storage in an optimistic manner.",
+						exchange.getExchangeId(), key);
+				try {
+					ByteArrayOutputStream bos = new ByteArrayOutputStream();
+					ObjectOutputStream oos = new ObjectOutputStream(bos);
+					oos.writeObject(holder);
+					oos.flush();
```
Comment (lburgazzoli) [inline components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java:340]: seems like serialize/deserialize of the exchange holder is quite common so you may think to extract the logic in some methods

File: components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+			ByteArrayOutputStream bos = new ByteArrayOutputStream();
+			ObjectOutputStream oos = new ObjectOutputStream(bos);
+			oos.writeObject(newHolder);
+			oos.flush();
+			byte[] data = bos.toByteArray();
+			CompletableFuture<PutResponse> completablePutResponse = kvClient.put(
+					ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()), ByteSequence.from(data));
+			completablePutResponse.get();
+		} catch (InterruptedException | ExecutionException | IOException e) {
+			LOG.error(e.getMessage(), e);
```
Comment (lburgazzoli) [inline components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java:187]: doesn't this case need to be properly handled ?

File: components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java (hunk 7 above)
Comment (malogan82) [reply components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java]: I think I should have fixed this issue, using transaction as you suggested, I've also fixed the other issues in Etcd3AggregationRepository class, meanwhile I'll fix indentation problems on xml files and add some test cases,
thanks

**Comment by malogan82** [comment]
sorry @lburgazzoli about this other one? thanks

**Comment by oscerd** [comment]
Can you please rebase?

**Comment by davsclaus** [comment]
Working on merging this

**Comment by davsclaus** [comment]
Thanks this has been merged.

@malogan82 you are welcome to help such as improving the docs
https://github.com/apache/camel/blob/master/components/camel-etcd3/src/main/docs/etcd3.adoc

Also the unit test can use test containers (docker) to run a local etcd3 and test against that (maybe)

**Comment by malogan82** [comment]
@davsclaus ok I'll improve docs and test, thanks
